drivers/platform/aducm3029/irq.c

```diff
@@ -142,34 +127,64 @@ int32_t irq_ctrl_init(struct irq_ctrl_desc **desc,
  */
 int32_t irq_ctrl_remove(struct irq_ctrl_desc *desc)
 {
+	uint32_t i;
+
 	if (!desc || !desc->extra || !initialized)
 		return FAILURE;
 
-	initialized = 0;
+	/* Free external interrupts */
+	for (i = 0; i < NB_EXT_INTERRUPTS; i++)
+		adi_xint_DisableIRQ(id_map_event[i]);
 	adi_xint_UnInit();
+
+	/* Free UART */
+	irq_unregister(desc, ADUCM_UART_INT_ID);
 	free(desc->extra);
 	free(desc);
+	initialized = 0;
 
 	return SUCCESS;
 }
 
 /**
- * @brief Register IRQ handling function for the specified <em>irq_id</em>.
- * @param desc - Interrupt controller descriptor.
- * @param irq_id - Id of the interrupt
- * @param callback_desc - Callback descriptor
- * @return \ref SUCCESS in case of success, \ref FAILURE otherwise.
+ * @brief Registers a IRQ callback function to irq controller.
+ * @param desc - The IRQ controller descriptor.
+ * @param irq_id - Interrupt identifier.
+ * @param callback_desc - Descriptor of the callback. If it is NULL, the
+ * callback will be unregistered
+ * @return SUCCESS in case of success, FAILURE otherwise.
  */
 int32_t irq_register_callback(struct irq_ctrl_desc *desc, uint32_t irq_id,
 			      struct callback_desc *callback_desc)
 {
-	if (!desc || !desc->extra || !initialized || !callback_desc ||
-	    irq_id >= NB_EXT_INTERRUPTS)
+	struct aducm_irq_ctrl_desc	*aducm_desc;
+	struct uart_desc		*uart_desc;
+
+	if (!desc || !desc->extra || !initialized ||  irq_id >= NB_INTERRUPTS)
 		return FAILURE;
 
-	((struct aducm_irq_ctrl_desc *)desc->extra)->callbacks[irq_id] =
-		*callback_desc;
+	if (!callback_desc)
+		return irq_unregister(desc, irq_id);
+
+	aducm_desc = desc->extra;
+
+	if (irq_id < NB_EXT_INTERRUPTS) {
+		aducm_desc->conf[irq_id].xint_conf =
+			(enum irq_mode)callback_desc->config;
+		adi_xint_RegisterCallback(id_map_event[irq_id],
+					  callback_desc->callback,
+					  callback_desc->ctx);
+	} else { //if (irq_id == ADUCM_UART_INT_ID) {
```

Comment:
We probably need this as an `else if` if we want to add more interrupt options to this. Also, this `if else` clause might get bigger if it will support all other drivers. Should we consider making it a `switch case` statement?

Reply:
Yes, a switch will be better